doc/mcxalter.1
.\" Copyright (c) 2011 Stijn van Dongen
.TH "mcxalter" 1 "21 Oct 2011" "mcxalter 11-294" "USER COMMANDS "
.po 2m
.de ZI
.\" Zoem Indent/Itemize macro I.
.br
'in +\\$1
.nr xa 0
.nr xa -\\$1
.nr xb \\$1
.nr xb -\\w'\\$2'
\h'|\\n(xau'\\$2\h'\\n(xbu'\\
..
.de ZJ
.br
.\" Zoem Indent/Itemize macro II.
'in +\\$1
'in +\\$2
.nr xa 0
.nr xa -\\$2
.nr xa -\\w'\\$3'
.nr xb \\$2
\h'|\\n(xau'\\$3\h'\\n(xbu'\\
..
.if n .ll -2m
.am SH
.ie n .in 4m
.el .in 8m
..
.SH NAME
mcxalter \- various network transformations
.SH SYNOPSIS

\fBmcxalter\fP
\fB[-imx\fP <fname> (\fIspecify matrix input\fP)\fB]\fP
\fB[-abc\fP <fname> (\fIspecify label input\fP)\fB]\fP
\fB[-tab\fP <fname> (\fIuse tab file\fP)\fB]\fP
\fB[-o\fP <fname> (\fIoutput\fP)\fB]\fP
\fB[-tf\fP spec (\fIapply tf-spec to input matrix\fP)\fB]\fP
.SH DESCRIPTION

This utility supplies various transformations of networks\&.
.SH OPTIONS

.ZI 2m "\fB-abc\fP <fname> (\fIlabel input\fP)"
\&
.br
The file name for input that is in label format\&.
.in -2m

.ZI 2m "\fB-imx\fP <fname> (\fIinput matrix\fP)"
\&
.br
The file name for input that is in mcl native matrix format\&.
.in -2m

.ZI 2m "\fB-tab\fP <fname> (\fIuse tab file\fP)"
\&
.br
This option causes the output to be printed with the labels
found in the tab file\&.
With \fB-abc\fP this option will, additionally, construct
a graph only on the labels found in the tab file\&.
If this option is used in conjunction with \fB-imx\fP the
tab domain and the matrix domain are required to be identical\&.
.in -2m

.ZI 2m "\fB-tf\fP <tf-spec> (\fItransform input matrix values\fP)"
\&
.br
Transform the input matrix values according
to the syntax described in \fBmcxio(5)\fP\&.
.in -2m

.ZI 2m "\fB-o\fP fname (\fIoutput file\fP)"
\&
.br
Output file\&.
.in -2m
.SH AUTHOR

Stijn van Dongen\&.
.SH SEE ALSO

\fBmcxio(5)\fP,
\fBmcx(1)\fP,
\fBmcxsubs(1)\fP,
and \fBmclfamily(7)\fP for an overview of all the documentation
and the utilities in the mcl family\&.


NEWS
$Id: NEWS,v 1.11 2003/11/29 21:30:05 flux Exp $

BLAST users may consider using the blast module. Add --enable-blast to
./configure and read the manual pages of mcxdeblast and mcxassemble.  and
optionally mclblastline. The blast module does not conflict with the tribe
module; you can install both of them simultaneously.

Read ChangeLog.


src/shmx/mcxalter.c
/*   (C) Copyright 2006, 2007, 2008 Stijn van Dongen
 *
 * This file is part of MCL.  You can redistribute and/or modify MCL under the
 * terms of the GNU General Public License; either version 3 of the License or
 * (at your option) any later version.  You should have received a copy of the
 * GPL along with MCL, in the file COPYING.
 *
 * NOTE
 * this file implements both /diameter/ and /ctty/ (centrality)
 * functionality for mcx.
*/

/* TODO
 * -  Handle components.
 * -  Clean up diameter/ctty organization.
*/


#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <time.h>

#include "mcx.h"

#include "util/types.h"
#include "util/ding.h"
#include "util/ting.h"
#include "util/io.h"
#include "util/err.h"
#include "util/opt.h"
#include "util/compile.h"

#include "impala/matrix.h"
#include "impala/tab.h"
#include "impala/io.h"
#include "impala/stream.h"
#include "impala/app.h"

static const char* me = "mcx alter";

enum
{  MY_OPT_ABC    =   MCX_DISP_UNUSED
,  MY_OPT_IMX
,  MY_OPT_TAB
,  MY_OPT_O
,  MY_OPT_CEILNB
,  MY_OPT_CANONICAL
}  ;


mcxOptAnchor alterOptions[] =
{  {  "-imx"
   ,  MCX_OPT_HASARG
   ,  MY_OPT_IMX
   ,  "<fname>"
   ,  "specify input matrix/graph"
   }
,  {  "-abc"
   ,  MCX_OPT_HASARG
   ,  MY_OPT_ABC
   ,  "<fname>"
   ,  "specify input using label pairs"
   }
,  {  "-tab"
   ,  MCX_OPT_HASARG
   ,  MY_OPT_TAB
   ,  "<fname>"
   ,  "specify tab file to be used with matrix input"
   }
,  {  "-o"
   ,  MCX_OPT_HASARG
   ,  MY_OPT_O
   ,  "<fname>"
   ,  "write to file fname"
   }
,  {  "-ceil-nb"
   ,  MCX_OPT_HASARG
   ,  MY_OPT_CEILNB
   ,  "<num>"
   ,  "remove edges from nodes with more than <num> edges"
   }
,  {  "-canonical"
   ,  MCX_OPT_HASARG | MCX_OPT_HIDDEN
   ,  MY_OPT_CANONICAL
   ,  "<num>"
   ,  "add canonical vector in the middle with <num> nodes"
   }
,  {  NULL, 0, 0, NULL, NULL }
}  ;


static  unsigned debug_g   =   -1;

static  mcxIO* xfmx_g      =   (void*) -1;
static  mcxIO* xfabc_g     =   (void*) -1;
static  mcxIO* xftab_g     =   (void*) -1;
static  mcxIO* xfout_g     =   (void*) -1;
static  mclTab* tab_g      =   (void*) -1;
static  dim ceilnb_g       =   -1;
static  dim canonical_g    =   -1;


static mcxstatus alterInit
(  void
)
   {  xfmx_g         =  mcxIOnew("-", "r")
   ;  xfabc_g        =  NULL
   ;  xftab_g        =  NULL
   ;  xfout_g        =  mcxIOnew("-", "w")
   ;  tab_g          =  NULL
   ;  ceilnb_g       =  0
   ;  canonical_g    =  0
   ;  return STATUS_OK
;  }


static mcxstatus alterArgHandle
(  int optid
,  const char* val
)
   {  switch(optid)
      {  case MY_OPT_IMX
      :  mcxIOnewName(xfmx_g, val)
      ;  break
      ;

         case MY_OPT_ABC
      :  xfabc_g = mcxIOnew(val, "r")
      ;  break
      ;

         case MY_OPT_O
      :  mcxIOnewName(xfout_g, val)
      ;  break
      ;

         case MY_OPT_CANONICAL
      :  canonical_g = atoi(val)
      ;  break
      ;  

         case MY_OPT_TAB
      :  xftab_g = mcxIOnew(val, "r")
      ;  break
      ;

         case MY_OPT_CEILNB
      :  ceilnb_g = atoi(val)
      ;  break
      ;

         default
      :  mcxExit(1)
      ;
   ;  }
      return STATUS_OK
;  }


static mcxstatus alterMain
(  int          argc_unused      cpl__unused
,  const char*  argv_unused[]    cpl__unused
)
   {  mclx* mx                =  NULL

   ;  mclxIOstreamer streamer = { 0 }

   ;  mcxLogLevel =
      MCX_LOG_AGGR | MCX_LOG_MODULE | MCX_LOG_IO | MCX_LOG_GAUGE | MCX_LOG_WARN
   ;  mclx_app_init(stderr)

   ;  if (xfabc_g)
      {  if (xftab_g)
            tab_g = mclTabRead(xftab_g, NULL, EXIT_ON_FAIL)
         ,  streamer.tab_sym_in = tab_g
      ;  mx
      =  mclxIOstreamIn
         (  xfabc_g
         ,     MCLXIO_STREAM_ABC
            |  MCLXIO_STREAM_MIRROR
            |  MCLXIO_STREAM_SYMMETRIC
            |  (tab_g ? MCLXIO_STREAM_GTAB_RESTRICT : 0)
         ,  NULL
         ,  mclpMergeMax
         ,  &streamer
         ,  EXIT_ON_FAIL
         )
      ;  tab_g = streamer.tab_sym_out
   ;  }
      else
      {  mx = mclxReadx(xfmx_g, EXIT_ON_FAIL, MCLX_REQUIRE_GRAPH)
      ;  if (xftab_g)
         tab_g = mclTabRead(xftab_g, mx->dom_cols, EXIT_ON_FAIL)
   ;  }

      mcxIOopen(xfout_g, EXIT_ON_FAIL)

   ;  if (ceilnb_g)
      {  dim n_hub, n_in, n_out
      ;  mclv* sel = mclgCeilNB(mx, ceilnb_g, &n_hub, &n_in, &n_out)
      ;  mcxLog
         (  MCX_LOG_FUNC
         ,  me
         ,  "considered %lu nodes in prepruning step (%lu needed, %lu edges out, %lu edges in)"
         ,  (ulong) sel->n_ivps
         ,  (ulong) n_hub
         ,  (ulong) n_out
         ,  (ulong) n_in
         )
      ;  mclvFree(&sel)
   ;  }

      if (canonical_g)
      {  dim d, e
      ;  for (d=1; d<=canonical_g; d++)
            e = d * (N_COLS(mx) / canonical_g)
         ,  mclvCanonical(mx->cols+e-1, e, 1.0)
   ;  }

      if (tab_g)
      {  mclxIOdumper dumper
      ;  mclxIOdumpSet(&dumper, MCLX_DUMP_PAIRS, NULL, NULL, NULL)
;fprintf(stderr, "HI\n")
      ;  if
         (  mclxIOdump
            (  mx
            ,  xfout_g
            ,  &dumper
            ,  tab_g
            ,  tab_g
            ,  MCLXIO_VALUE_GETENV
            ,  RETURN_ON_FAIL
          ) )
         mcxDie(1, me, "something suboptimal")
   ;  }
      else
      {  if (mclxWrite(mx, xfout_g, MCLXIO_VALUE_GETENV, RETURN_ON_FAIL))
         mcxDie(1, me, "something less than perfect. AAAAAARGH!")
   ;  }

      return 0
;  }


mcxDispHook* mcxDispHookAlter
(  void
)
   {  static mcxDispHook alterEntry
   =  {  "alter"
      ,  "alter [options]"
      ,  alterOptions
      ,  sizeof(alterOptions)/sizeof(mcxOptAnchor) - 1

      ,  alterArgHandle
      ,  alterInit
      ,  alterMain

      ,  0
      ,  0
      ,  MCX_DISP_DEFAULT
      }
   ;  return &alterEntry
;  }


